plugins/woocommerce/includes/data-stores/class-wc-product-data-store-cpt.php

```diff
@@ -145,8 +145,17 @@ private function obtain_lock_on_sku_for_concurrent_requests( $product ) {
 			return boolval( $locked );
 		}
 
-		// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
-		$result = $wpdb->query( $query );
+		// The insert query can potentially result in a deadlock if there is high concurrency
+		// when trying to insert products, which will result in a false negative for SKU lock
+		// and incorrectly products not being created.
+		// To mitigate this, we will retry the query 3 times before giving up.
+		for ( $attempts = 0; $attempts < 3; $attempts++ ) {
+			// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
```

Comment:
When `$attempts > 0` should we introduce a very short delay or, if not, should we provide the opportunity for deployments to do so via an action hook?

I'm not sure this is necessary—we could definitely start here and see what further feedback comes in—but wanted to pause and ask.

Reply:
What about we start adding that short delay after the first attempt and see how it goes:
```php
for ( $attempts = 0; $attempts < 3; $attempts++ ) {
	if ( $attempts > 1 ) {
		usleep( 10000 );
	}

	// phpcs:ignore WordPress.DB.PreparedSQL.NotPrepared
	$result = $wpdb->query( $query );
	if ( false !== $result ) {
		break;
	}
}
```
And later, if necessary, we can think about adding a hook? 🤔

Reply:
That sounds great, and 10ms is probably a pretty reasonable default.